src/wp-includes/block-template-utils.php

```diff
@@ -360,6 +360,17 @@ function _get_block_templates_files( $template_type, $query = array() ) {
 				continue;
 			}
 
+			// The child theme items (stylesheet) are processed before the parent theme's (template).
+			// If a child theme defines a template, prevent the parent template from being added to the list as well.
+			$slug_already_present = false !== array_search(
+				$template_slug,
+				wp_list_pluck( $template_files, 'slug' ),
+				true
+			);
+			if ( $slug_already_present ) {
+				continue;
+			}
```

Comment:
This could be implemented in a more efficient way, with both `array_search()` and `wp_list_pluck()` operating on the entire array being rather expensive from a PHP logic perspective.

The following would be much faster:
* Change `$template_files` to be a map of `$slug => $candidate`, rather than indexed array of `$candidate` entries.
* Use `isset( $template_files[ $slug ] )` here instead for the early continue.
* At the end of the function return `array_values( $template_files )` to maintain the original return type.

Reply:
Implemented at https://github.com/WordPress/wordpress-develop/pull/4940/commits/511cc7203b4c294e4e1dfd3f2e2e9f28958df2f9